src/ra_server.erl

```diff
@@ -522,58 +522,61 @@ handle_leader({PeerId, #append_entries_reply{success = false}},
-                         cluster := Nodes, log := Log0}) ->
+                                             next_index = PeerNextIdx,
+                                             last_index = PeerLastIdx,
+                                             last_term = PEER_LAST_TERM}},
+              #{cfg := #cfg{log_id = LogId} = Cfg,
+                cluster := Nodes, log := Log0} = State0) ->
     ok = incr_counter(Cfg, ?C_RA_SRV_AER_REPLIES_FAILED, 1),
-    #{PeerId := Peer0 = #{match_index := MI,
-                          next_index := NI}} = Nodes,
+    #{PeerId := #{match_index := MI,
+                  next_index := NI} = Peer0} = Nodes,
     % if the last_index exists and has a matching term we can forward
     % match_index and update next_index directly
-    {Peer, Log} = case ra_log:fetch_term(LastIdx, Log0) of
-                      {undefined, L} ->
-                          % entry was not found - simply set next index to
-                          ?DEBUG("~ts: setting next index for ~w ~b",
-                                 [LogId, PeerId, NextIdx]),
-                          %% TODO: match index should not be set here surely??
-                          {Peer0#{match_index => LastIdx,
-                                  next_index => NextIdx}, L};
+    {Peer, Log} = case ra_log:fetch_term(PeerLastIdx, Log0) of
+                      {undefined, Log1} ->
+                          % entry was not found
+                          ?DEBUG("~ts: peer ~w last index ~b not found "
+                                 "setting next index to ~b",
+                                 [LogId, PeerId, PeerLastIdx, PeerNextIdx]),
+                          {Peer0#{next_index => PeerNextIdx}, Log1};
                       % entry exists we can forward
-                      {LastTerm, L} when LastIdx >= MI ->
+                      {PEER_LAST_TERM = Term, Log1}
+                        when is_integer(Term) andalso
+                             PeerLastIdx >= MI ->
                           ?DEBUG("~ts: setting last index to ~b, "
                                  " next_index ~b for ~w",
-                                 [LogId, LastIdx, NextIdx, PeerId]),
-                          {Peer0#{match_index => LastIdx,
-                                  next_index => NextIdx}, L};
-                      {_Term, L} when LastIdx < MI ->
+                                 [LogId, PeerLastIdx, PeerNextIdx, PeerId]),
+                          {Peer0#{match_index => PeerLastIdx,
+                                  next_index => PeerNextIdx}, Log1};
+                      {_OtherTerm, Log1}
+                        when PeerLastIdx < MI ->
                           % TODO: this can only really happen when peers are
                           % non-persistent.
-                          % should they turn-into non-voters when this sitution
-                          % is detected
+                          % Should they turn-into non-voters when this sitution
```

Comment:
```suggestion
                          % Should they turn-into non-voters when this situation
```

might as well fix this since the line is already being updated